File: go/Day13.go
package main

import (
	"aoc2021/Utils"
	"fmt"
	"strings"
)

func Day13_Part1() {
	paper := make([][]bool, 0)
	lines := Utils.ReadFile("../data/day13.txt")
	var folding_section_start = 0

	for _, line := range lines {
		folding_section_start++

		if line == "" {
			break
		}

		var x = Utils.StrToInt(strings.Split(line, ",")[0], 10)
		var y = Utils.StrToInt(strings.Split(line, ",")[1], 10)

		for len(paper) <= y {
			paper = append(paper, make([]bool, 0))
		}
		for len(paper[y]) <= x {
			paper[y] = append(paper[y], false)
		}

		paper[y][x] = true
	}

	var longest = 0
	for _, line := range paper {
		if len(line) > longest {
			longest = len(line)
		}
	}

	for line_id := range paper {
		for len(paper[line_id]) < longest {
			paper[line_id] = append(paper[line_id], false)
		}
	}

	var fold_line = lines[folding_section_start]
	var fl_splitted = strings.Split(fold_line, " ")
	var fold_index = Utils.StrToInt(strings.Split(fl_splitted[len(fl_splitted)-1], "=")[1], 10)

	for line := range paper {
		for index := 0; index < longest-fold_index-1; index++ {
			paper[line][index] = paper[line][index] || paper[line][longest-index-1]
		}
		for remove := 0; remove <= fold_index; remove++ {
			paper[line] = Utils.RemoveAtBool(paper[line], fold_index)
		}
	}

	var result = 0
	for _, line := range paper {
		for _, char := range line {
			if char {
				result++
			}
		}
	}

	fmt.Printf("Day: 13 | Part: 1 | Result: %d\n", result)
}

func Day13_Part2() {
	paper := make([][]bool, 0)
	lines := Utils.ReadFile("../data/day13.txt")
	var folding_section_start = 0

	for _, line := range lines {
		folding_section_start++

		if line == "" {
			break
		}

		var x = Utils.StrToInt(strings.Split(line, ",")[0], 10)
		var y = Utils.StrToInt(strings.Split(line, ",")[1], 10)

		for len(paper) <= y {
			paper = append(paper, make([]bool, 0))
		}
		for len(paper[y]) <= x {
			paper[y] = append(paper[y], false)
		}

		paper[y][x] = true
	}

	var longest = 0
	for _, line := range paper {
		if len(line) > longest {
			longest = len(line)
		}
	}

	for line_id := range paper {
		for len(paper[line_id]) < longest {
			paper[line_id] = append(paper[line_id], false)
		}
	}

	for fold_line_id := folding_section_start; fold_line_id < len(lines); fold_line_id++ {
		longest = 0
		for _, line := range paper {
			if len(line) > longest {
				longest = len(line)
			}
		}

		var fl_splitted = strings.Split(lines[fold_line_id], " ")
		var fold_line = fl_splitted[len(fl_splitted)-1]
		var fold_axis = strings.Split(fold_line, "=")[0]
		var fold_index = Utils.StrToInt(strings.Split(fold_line, "=")[1], 10)

		if fold_axis == "x" {
			for line := range paper {
				for index := 0; index < longest-fold_index-1; index++ {
					paper[line][index] = paper[line][index] || paper[line][longest-index-1]
				}
				for remove := 0; remove <= fold_index; remove++ {
					paper[line] = Utils.RemoveAtBool(paper[line], fold_index)
				}
			}
		} else if fold_axis == "y" {
			for line := 0; line < len(paper)-fold_index-1; line++ {
				for index := 0; index < longest; index++ {
					paper[line][index] = paper[line][index] || paper[len(paper)-line-1][index]
				}
			}
			for remove := 0; remove <= fold_index; remove++ {
				paper = Utils.RemoveAtArrayBool(paper, fold_index)
			}
		}
	}

	var result = 0
	for _, line := range paper {
		for _, char := range line {
			if char {
				result++
			}
		}
	}

	fmt.Println("Day: 13 | Part: 1 | Result:")
	for _, line := range paper {
		for _, c := range line {
			var char = " "
			if c {
				char = "#"
			}

			fmt.Printf("%s", char)
		}
		fmt.Println()
	}
}

File: java/src/cz/simik31/aoc2021/Day13.java
package cz.simik31.aoc2021;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Day13 {
    public static void part_1() {
        List<List<Boolean>> paper = new ArrayList<>();
        List<String> lines = Utils.readFileLines("../data/day13.txt");
        int folding_section_start = 0;

        for (String s : lines) {
            folding_section_start++;
            if (s.equals("")) break;

            List<Integer> ln = Arrays.stream(s.split(",")).map(Integer::parseInt).toList();
            int x = ln.get(0), y = ln.get(1);

            while (paper.size() <= y) paper.add(new ArrayList<>());
            while (paper.get(y).size() <= x) paper.get(y).add(false);

            paper.get(y).set(x, true);
        }

        int longest = 0;
        for (List<Boolean> line : paper) longest = Math.max(longest, line.size());

        for (List<Boolean> line : paper) while (line.size() < longest) line.add(false);

        String fold_line = lines.get(folding_section_start);
        int fold_index = Integer.parseInt(fold_line.substring(fold_line.lastIndexOf(" ")).split("=")[1]);

        for (List<Boolean> line : paper) {
            for (int index = 0; index < longest - fold_index - 1; index++)
                line.set(index, line.get(index) | line.get(longest - index - 1));

            for (int remove = 0; remove <= fold_index; remove++)
                line.remove(fold_index);
        }

        int result = 0;
        for (List<Boolean> line : paper)
            for (boolean c : line)
                if (c) result++;

        System.out.println("Day: 13 | Part: 1 | Result: " + result);
    }

    public static void part_2() {
        List<List<Boolean>> paper = new ArrayList<>();
        List<String> lines = Utils.readFileLines("../data/day13.txt");
        int folding_section_start = 0;

        for (String s : lines) {
            folding_section_start++;
            if (s.equals("")) break;

            List<Integer> ln = Arrays.stream(s.split(",")).map(Integer::parseInt).toList();
            int x = ln.get(0), y = ln.get(1);

            while (paper.size() <= y) paper.add(new ArrayList<>());
            while (paper.get(y).size() <= x) paper.get(y).add(false);

            paper.get(y).set(x, true);
        }

        int longest = 0;
        for (List<Boolean> line : paper) longest = Math.max(longest, line.size());

        for (List<Boolean> line : paper) while (line.size() < longest) line.add(false);

        for (int fold_line_id = folding_section_start; fold_line_id < lines.size(); fold_line_id++) {
            longest = 0;
            for (List<Boolean> line : paper) longest = Math.max(longest, line.size());

            String fold_line = lines.get(fold_line_id);
            String[] interest = fold_line.substring(fold_line.lastIndexOf(" ") + 1).split("=");
            String fold_axis = interest[0];
            int fold_index = Integer.parseInt(interest[1]);

            if (fold_axis.equals("x")) {
                for (List<Boolean> line : paper) {
                    for (int index = 0; index < longest - fold_index - 1; index++)
                        line.set(index, line.get(index) | line.get(longest - index - 1));

                    for (int remove = 0; remove <= fold_index; remove++)
                        line.remove(fold_index);
                }
            }
            else if (fold_axis.equals("y")) {
                for (int line = 0; line < paper.size() - fold_index - 1; line++)
                    for (int index = 0; index < paper.get(line).size(); index++)
                        paper.get(line).set(index, paper.get(line).get(index) | paper.get(paper.size() - line - 1).get(index));

                for (int remove = 0; remove <= fold_index; remove++)
                    paper.remove(fold_index);
            }
        }

        System.out.println("Day: 13 | Part: 1 | Result:");

        for (List<Boolean> line : paper) {
            for (boolean c : line)
                System.out.print(c ? "#" : " ");
            System.out.println();
        }

    }
}
